syntax = "proto3";

package fleet;

option go_package = "github.com/amr-saas/gateway/proto";

// Fleet Gateway Service
// Backend ↔ Gateway間の双方向通信を提供
service FleetGateway {
  // ロボット状態のストリーミング（Server Streaming）
  // Backendがsubscribeし、Gatewayがロボット状態をpush
  rpc StreamRobotStatus(StreamRequest) returns (stream RobotStatus);
  
  // 全ロボット状態の取得（Unary）
  rpc GetAllRobots(Empty) returns (RobotList);
  
  // 特定ロボットの状態取得（Unary）
  rpc GetRobot(RobotId) returns (RobotStatus);
  
  // ロボットへのコマンド送信（Unary）
  rpc SendCommand(Command) returns (CommandResponse);
  
  // ミッションの開始（Unary）
  rpc StartMission(MissionRequest) returns (MissionResponse);
  
  // ミッションのキャンセル（Unary）
  rpc CancelMission(MissionId) returns (MissionResponse);
  
  // Gatewayのヘルスチェック（Unary）
  rpc HealthCheck(Empty) returns (HealthResponse);
}

// 共通メッセージ
message Empty {}

message RobotId {
  string id = 1;
}

message MissionId {
  string id = 1;
}

// ストリーミング購読リクエスト
message StreamRequest {
  repeated string robot_ids = 1;  // 空の場合は全ロボット
  int32 interval_ms = 2;          // 更新間隔（ミリ秒）
}

// 位置情報
message Position {
  double x = 1;
  double y = 2;
  double theta = 3;  // 向き（ラジアン）
}

// ロボット状態
message RobotStatus {
  string id = 1;
  string name = 2;
  string vendor = 3;
  string model = 4;
  RobotState state = 5;
  Position position = 6;
  int32 battery_level = 7;        // 0-100
  string current_mission_id = 8;  // 空文字の場合はミッションなし
  int64 last_seen = 9;            // Unix timestamp（ミリ秒）
  map<string, string> metadata = 10;
}

// ロボット状態列挙
enum RobotState {
  UNKNOWN = 0;
  IDLE = 1;
  MOVING = 2;
  PAUSED = 3;
  ERROR = 4;
  CHARGING = 5;
  OFFLINE = 6;
}

// ロボットリスト
message RobotList {
  repeated RobotStatus robots = 1;
  int32 total = 2;
}

// コマンド
message Command {
  string robot_id = 1;
  CommandType type = 2;
  map<string, string> parameters = 3;
}

enum CommandType {
  CMD_UNKNOWN = 0;
  CMD_MOVE = 1;           // 移動
  CMD_STOP = 2;           // 停止
  CMD_PAUSE = 3;          // 一時停止
  CMD_RESUME = 4;         // 再開
  CMD_DOCK = 5;           // 充電ステーションへ移動
  CMD_UNDOCK = 6;         // 充電ステーションから離脱
  CMD_LOCALIZE = 7;       // 自己位置推定
}

// コマンドレスポンス
message CommandResponse {
  bool success = 1;
  string message = 2;
  string command_id = 3;  // トレース用ID
}

// ミッションリクエスト
message MissionRequest {
  string mission_id = 1;
  string robot_id = 2;
  MissionType type = 3;
  repeated Waypoint waypoints = 4;
  int32 priority = 5;
  map<string, string> parameters = 6;
}

enum MissionType {
  MISSION_UNKNOWN = 0;
  MISSION_TRANSPORT = 1;      // 搬送
  MISSION_PATROL = 2;         // 巡回
  MISSION_CHARGE = 3;         // 充電
  MISSION_CUSTOM = 4;         // カスタム
}

// ウェイポイント
message Waypoint {
  string id = 1;
  Position position = 2;
  WaypointAction action = 3;
  int32 wait_time_sec = 4;
}

enum WaypointAction {
  WP_PASS = 0;       // 通過
  WP_STOP = 1;       // 停止
  WP_PICKUP = 2;     // ピックアップ
  WP_DROPOFF = 3;    // ドロップオフ
}

// ミッションレスポンス
message MissionResponse {
  bool success = 1;
  string message = 2;
  MissionStatus status = 3;
}

enum MissionStatus {
  MISSION_STATUS_UNKNOWN = 0;
  MISSION_PENDING = 1;
  MISSION_IN_PROGRESS = 2;
  MISSION_COMPLETED = 3;
  MISSION_CANCELLED = 4;
  MISSION_FAILED = 5;
}

// ヘルスレスポンス
message HealthResponse {
  bool healthy = 1;
  string version = 2;
  int32 connected_robots = 3;
  int64 uptime_seconds = 4;
}

// ===========================================
// Data Recording Service (Gateway -> Backend)
// ===========================================

// センサデータ記録サービス
service DataRecordingService {
  // バッチでセンサデータを記録
  rpc RecordSensorData(BatchSensorDataRequest) returns (BatchSensorDataResponse);
  
  // ストリーミングでセンサデータを記録
  rpc StreamSensorData(stream SensorDataRecord) returns (BatchSensorDataResponse);
}

// センサデータレコード
message SensorDataRecord {
  string robot_id = 1;
  int64 timestamp = 2;
  map<string, double> sensor_data = 3;   // センサ値 (e.g., lidar, imu, encoder)
  map<string, double> control_data = 4;  // 制御値 (e.g., velocity, steering)
}

// バッチセンサデータリクエスト
message BatchSensorDataRequest {
  repeated SensorDataRecord records = 1;
}

// バッチセンサデータレスポンス
message BatchSensorDataResponse {
  bool success = 1;
  int32 recorded_count = 2;
  string message = 3;
}
